### PR TITLE
Got CI working again (Correctly identifies everything that is failing)

### PR DESCRIPTION
**Pull request by torch2424**
This PR finalizes getting all of the CI to work (in terms of catching failing tests).

These fixes come from mostly checking the build logs in: https://github.com/DuncanUszkay1/assemblyscript/runs/846787874

This will be organized by the things that are fixed

## Features testing

See: https://github.com/DuncanUszkay1/assemblyscript/runs/846789583

This is fixed by recreating `tests/compiler/features/reference-types.untouched.wat`. Which is now updated to use closures.

## Loader Testing

See: https://github.com/DuncanUszkay1/assemblyscript/runs/846789746

This is actually working correctly, with an actionable error. It has to do with exporting a closure: https://github.com/DuncanUszkay1/assemblyscript/blob/master/lib/loader/tests/assembly/index.ts#L62

## Bootstrap Testing

See: https://github.com/DuncanUszkay1/assemblyscript/runs/846789713

This is also working correctly, but finds a bug that I started fixing, that I think may be worth including in this PR. 

This attempts to do a self-compilation of AssemblyScript, and works for most of the program, but breaks on a recursive closure you can find here: https://github.com/DuncanUszkay1/assemblyscript/blob/master/src/flow.ts#L1398

And some screenshots to prove it:

Before fix:
Screenshot from 2020-07-08 16-26-31

After fixing the name conflict:
Screenshot from 2020-07-08 16-23-15

Still broken:
Screenshot from 2020-07-08 16-22-50

## Hanging tests on validation:

This is apparent as of: https://github.com/DuncanUszkay1/assemblyscript/runs/846789647 Which hung for 5 hours :joy: 

The fix is in this PR at: `cli/asc.js` where a return is incorectly placed in a callback. This fix should be reflected in AS main as well, and I'll do that tomorrow :smile: 

EDIT: Opened here: https://github.com/AssemblyScript/assemblyscript/pull/1383#issue-447007257

## In conclusion

And this leaves us with this remaining CI result:

Screenshot from 2020-07-08 17-52-24

So now we have a much more clear guide on what/how things need to be fixed in closures to get tests passing again :smile: :+1:

### CODE REVIEW

**Comment by torch2424** [comment]
Also, I went ahead and commited re-creating all the test `.wat` since that is compared in the tests and will fail if it doesnt match (which is what we expect since we are changing this :smile: )

**Comment by torch2424** [comment]
@DuncanUszkay1 I went ahead and opened the PR for the hanging tests on AS Main here: https://github.com/AssemblyScript/assemblyscript/pull/1383#issue-447007257

I still think we should merge this though, since we'll have to sync with AS eventually anyways, and if the PR gets merged as-is, shouldn't cause any issues :smile:

**Comment by DuncanUszkay1** [comment]
Merge this PR? Seems like the stated goal of the PR isn't done yet given the failing tests though- or am I confused?

**Comment by torch2424** [comment]
@DuncanUszkay1 

> Merge this PR? Seems like the stated goal of the PR isn't done yet given the failing tests though- or am I confused?

Oh yes! We can merge! The goal of this PR was to get the tests running in full again (instead of failing / hanging early on) :smile: In previous builds the tests weren't actually running due to the various reasons that I fixed in previous PRs and this one. 

This PR now shows we are currently failing 6 compiler tests, so next would be to open a PR to start fixing our closures implementation so we start passing more tests :smile:
